contrib/scalar/scalar.c

```diff
@@ -0,0 +1,391 @@
+	    git_config_set_in_file_gently(config_path,
+					  "remote.origin.fetch",
+					  /*
+					   * TODO: should we respect
+					   * single_branch here?
+					   */
+					  "+refs/heads/*:refs/remotes/origin/*") ||
+	    git_config_set_in_file_gently(config_path,
+					  "remote.origin.promisor", "true") ||
+	    git_config_set_in_file_gently(config_path,
+					  "remote.origin.partialCloneFilter",
+					  "blob:none"))
+		return error(_("could not configure '%s'"), dir);
+
+	if (!full_clone &&
+	    (res = run_git(dir, "-c", "core.useGVFSHelper=false",
+			   "sparse-checkout", "init", "--cone", NULL)))
+		goto cleanup;
+
+	if (set_recommended_config(config_path))
+		return error(_("could not configure '%s'"), dir);
+
+	/*
+	 * TODO: should we pipe the output and grep for "filtering not
+	 * recognized by server", and suppress the error output in
+	 * that case?
+	 */
+	if ((res = run_git(dir, "-c", "core.useGVFSHelper=false", "fetch",
+			   "--quiet", "origin", NULL))) {
+		warning(_("Partial clone failed; Trying full clone"));
+
+		if (git_config_set_in_file_gently(config_path,
+						  "remote.origin.promisor",
+						  NULL) ||
+		    git_config_set_in_file_gently(config_path,
+						  "remote.origin.partialCloneFilter",
+						  NULL)) {
+			res = error(_("could not configure for full clone"));
+			strvec_clear(&args);
+			goto cleanup;
+		}
+
+		if ((res = run_git(dir, "-c", "core.useGVFSHelper=false",
+				   "fetch", "--quiet", "origin", NULL)))
+			goto cleanup;
+	}
+
+
+	die("To be continued");
+
+cleanup:
+	free(dir);
+	strbuf_release(&buf);
+	return res;
+}
+
+static int cmd_config(int argc, const char **argv)
+{
+	die(N_("'%s' not yet implemented"), argv[0]);
+}
+
+static int cmd_diagnose(int argc, const char **argv)
+{
+	die(N_("'%s' not yet implemented"), argv[0]);
+}
+
+static int cmd_list(int argc, const char **argv)
+{
+	return run_git(NULL, "config", "--get-all", "scalar.repo", NULL);
+}
+
+static int add_or_remove_enlistment(int add)
+{
+	int res;
+
+	if (!the_repository->worktree)
+		die(_("Scalar enlistments require a worktree"));
+
+	res = run_git(NULL, "config", "--global", "--get",
+		      "--fixed-value", "scalar.repo", the_repository->worktree, NULL);
+
+	/*
+	 * If we want to add and the setting is already there, then do nothing.
+	 * If we want to remove and the setting is not there, then do nothing.
+	 */
+	if ((add && !res) || (!add && res))
+		return 0;
+
+	return run_git(NULL, "config", "--global",
+		       add ? "--add" : "--unset",
+		       "--fixed-value", "scalar.repo",
+		       the_repository->worktree, NULL);
+}
+
+static int initialize_enlistment_id(const char *file)
+{
+	const char *key = "scalar.enlistment-id";
+	char *value;
+	struct strbuf id = STRBUF_INIT;
+	int res;
+
+	if (!file && !git_config_get_string(key, &value)) {
+		trace2_data_string("scalar", the_repository, "enlistment-id", value);
+		free(value);
+		return 0;
+	}
+
+	strbuf_addstr(&id, "TODO:GENERATE-GUID");
```

Comment:
I have to wonder if this needs to be an actual GUID / UUID (or if we even still need it).
Linking in GUID / UUID code can be a pain in the Makefile.
I'm wondering if a hash value (of say the current time) would be sufficient here?

Reply:
I was thinking something like that, but if we could get a machine name or something, then that could help make it more unique.

Reply:
How about a hash of `git_committer_info(0)` with `getnanotime()` appended? That should be unique enough.